pymeasure/instruments/instrument.py

```diff
@@ -149,134 +257,200 @@ def binary_values(self, command, header_bytes=0, dtype=np.float32):
+                    else:
+                        raise ValueError(
+                            'Values of type `{}` are not allowed '
+                            'for Instrument.control'.format(type(values))
+                        )
+                else:
+                    vals = get_process(vals)
+                    return vals
 
-        def fget(self):
-            vals = self.values(get_command, **kwargs)
-            if check_get_errors:
-                self.check_errors()
-            if len(vals) == 1:
-                value = get_process(vals[0])
+        if set_command is None:
+            def fset(self, value):
+                raise LookupError("Instrument property can not be set.")
+        else:
+            def fset(self,
+                     value,
+                     set_command=set_command,
+                     validator=validator,
+                     values=values,
+                     map_values=map_values,
+                     set_process=set_process,
+                     check_set_errors=check_set_errors,
+                     ):
+
+                value = set_process(validator(value, values))
                 if not map_values:
-                    return value
+                    pass
                 elif isinstance(values, (list, tuple, range)):
-                    return values[int(value)]
+                    value = values.index(value)
                 elif isinstance(values, dict):
-                    return inverse[value]
+                    value = values[value]
                 else:
                     raise ValueError(
                         'Values of type `{}` are not allowed '
                         'for Instrument.control'.format(type(values))
                     )
-            else:
-                vals = get_process(vals)
-                return vals
-
-        def fset(self, value):
-            value = set_process(validator(value, values))
-            if not map_values:
-                pass
-            elif isinstance(values, (list, tuple, range)):
-                value = values.index(value)
-            elif isinstance(values, dict):
-                value = values[value]
-            else:
-                raise ValueError(
-                    'Values of type `{}` are not allowed '
-                    'for Instrument.control'.format(type(values))
-                )
-            self.write(set_command % value)
-            if check_set_errors:
-                self.check_errors()
+                self.write(set_command % value)
+                if check_set_errors:
+                    self.check_errors()
 
         # Add the specified document string to the getter
         fget.__doc__ = docs
 
-        return property(fget, fset)
+        if dynamic:
+            # Compute list of parameters supporting dynamic value
+            fget_list = []
+            for var in fget.__code__.co_varnames:
```

Comment:
I know Python is dynamic and all, but do we need to do this inspect-like dynamic finding of the arguments? We _know_ the admissible set of arguments because fget/fset are under our control here, so co_varnames will always be the same afaict. Further, we already use hard-coded values in Instrument.special_keys, anyway. 
Maybe easier to have two hardcoded special_keys lists for get and set, and those are at the same time fget_list/fset_list? Thinking further, maybe this makes fget_list/fset_list wholly unnecessary as we have those as module-level constants.
The only thing we maybe need is an additional `**kwargs` in the error-raising fget/fset above (if we don't use None instead)

Reply:
@bilderbuchi, I think this is a very valuable feedback which can potentially simplify a lot. Let me review it and I will get back to you. In all cases, the obscure `fget.__code__.co_varnames` should go.

Reply:
@bilderbuchi, I have made some code simplification following your comments and I did something different which I would like to have your comment.

In particular, I have removed all the module global variables to have DynamicProperty class more general and not specific for Instrument class, even though I have no example of other use for it )also unit testing should be easier if implemented in future). 
Very important: after make the modification you suggested, I am proposing also to remove the dynamic parameter and have all the Instrument.control/measurement/setting properties being dynamic.
I left the documentation around in case this change does not find consensus, so I can revert back it.

Could You please have a look and send your feedback at your convenience ?

Reply:
This looks so much better now! We have already clarified that `dynamic` should remain. 
I see now why you keep the fset_params_list in Instrument, not DynamicProperty, so my recent comment to that effect is not valid -- I'll strike that out.